libraries/Microsoft.Bot.Connector/Authentication/CloudEnvironment.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Net.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    /// <summary>
+    /// Cloud environments capture the environment specific Bot Framework Protocol auth code.
+    /// </summary>
+    public static class CloudEnvironment
+    {
+        /// <summary>
+        /// Creates the appropriate cloud environment instance.
+        /// </summary>
+        /// <param name="channelService">The Channel Service.</param>
+        /// <param name="validateAuthority">The validate authority value to use.</param>
+        /// <param name="toChannelFromBotLoginUrl">The to Channel from bot login url.</param>
+        /// <param name="toChannelFromBotOAuthScope">The to Channel from bot oauth scope.</param>
+        /// <param name="toBotFromChannelTokenIssuer">The to bot from Channel Token Issuer.</param>
+        /// <param name="oAuthUrl">The oAuth url.</param>
+        /// <param name="toBotFromChannelOpenIdMetadataUrl">The to bot from Channel Open Id Metadata url.</param>
+        /// <param name="toBotFromEmulatorOpenIdMetadataUrl">The to bot from Emulator Open Id Metadata url.</param>
+        /// <param name="callerId">The Microsoft app password.</param>
+        /// <param name="credentialFactory">The IServiceClientCredentialsFactory to use to create credentials.</param>
+        /// <param name="authConfiguration">The AuthenticationConfiguration to use.</param>
+        /// <param name="httpClient">The HttpClient to use.</param>
+        /// <param name="logger">The ILogger instance to use.</param>
+        /// <returns>A new cloud environment.</returns>
+        public static ICloudEnvironment Create(
+            string channelService,
```

Comment:
From usage, there seem to be 2 practical ways of calling this: Just pass the last 4 parameters, or pass all parameters. Should we have an additional overload to represent both scenarios? The call where we pass all the nulls feels like it could be improved. Not critical, but a nice usability consideration potentially.

Reply:
yes I think something like that should work - see previous comment - I think we can add this.

to be honest I think we have way to many ways to create things and way to many overloads in general - so I wanted to start with the "worst case" and then add overloads for common usage.

for example, private functions with default arguments just seems like a waste of time to me - and often not helpful for reading the code